swayidle/main.c

```diff
@@ -368,11 +369,17 @@ static int display_event(int fd, uint32_t mask, void *data) {
 	if (mask & WL_EVENT_HANGUP) {
 		sway_terminate(0);
 	}
-	if (wl_display_dispatch(state.display) < 0) {
-		wlr_log_errno(WLR_ERROR, "wl_display_dispatch failed, exiting");
-		sway_terminate(0);
+	while (wl_display_prepare_read(state.display) != 0) {
```

Comment:
This needs to be called before polling, ie. in the main loop. Docs say:

>Use this function before polling on the display fd

Reply:
(See https://people.freedesktop.org/~whot/wayland-doxygen/wayland/Client/classwl__display.html#a40039c1169b153269a3dc0796a54ddb0 for a complete example)

Reply:
So move the `prepare_read` to the main loop? But then how would we know whether to `cancel_read`? (which I forgot to put in my attempt either, but eh)

Reply:
Also, why doesn't this use the default event loop?

Reply:
>But then how would we know whether to cancel_read? 

We could probably have a `bool read_wl_events` variable set to `false` before dispatching, set it to `true` if we read from the Wayland socket, and call `wl_display_cancel_read` after disptaching if `read_wl_events` is still `false`.

>Also, why doesn't this use the default event loop?

Which default event loop?

We use the event loop from libwayland-server, but the `wl_display` doesn't have an event loop because we're a client. We should probably migrate to our own event loop instead.

Reply:
I've done that now.
And I was thinking about the default event queue, sorry.